### ISSUE TITLE
Memory usage on rserve AND rockfiles increases running release-test.R 

### ISSUE THREAD

**Issue by clemens-tolboom** [molgenis/molgenis-service-armadillo#529]
# Testing

**Note: `rserve` images behave similar**
 
Running the `release-test.R` script ie 10 times against armadillo server the docker containers memory usage increase from ~100MB to ~ 3GB

## Causes

- our script does weird things?
  - Running ie https://github.com/timcadman/ds-helper/tree/dev/vignettes shows similar memory increase
- are data files not freed
- is it `resourcer` package
- ?

## Docker containers

This seems not to depend on either rserve or rock based images.

```
docker container list
CONTAINER ID   IMAGE                                              COMMAND                   CREATED          STATUS          PORTS                                       NAMES
29c65e2733ed   datashield/rock-base:latest                        "/docker-entrypoint.…"    25 minutes ago   Up 25 minutes   0.0.0.0:8085->8085/tcp, :::8085->8085/tcp   rock
28fee48632b2   datashield/armadillo-rserver_caravan-xenon:1.0.0   "/bin/sh -c 'R -e \"M…"   26 minutes ago   Up 26 minutes   0.0.0.0:6313->6311/tcp, :::6313->6311/tcp   xenon
fc1ceb3febb8   datashield/armadillo-rserver                       "/bin/sh -c 'R -e \"M…"   26 minutes ago   Up 26 minutes   0.0.0.0:6311->6311/tcp, :::6311->6311/tcp   default
```
### First 10 lines

```
root@armadillo-backend:~# head -10 server-log-script.log 

TIME                CONTAINER ID   NAME      CPU %     MEM USAGE / LIMIT     MEM %     NET I/O          BLOCK I/O        PIDS
2023-10-26T14:56:54                --        0.00%     0B / 0B             0.00%     0B / 0B   0B / 0B     0
2023-10-26T14:56:54 fc1ceb3febb8   default   0.01%     83.98MiB / 7.744GiB   1.06%     736B / 0B   18.3MB / 0B   6
2023-10-26T14:56:56 fc1ceb3febb8   default   0.01%     83.98MiB / 7.744GiB   1.06%     736B / 0B   18.3MB / 0B   6
2023-10-26T14:56:58 fc1ceb3febb8   default   0.01%     83.98MiB / 7.744GiB   1.06%     806B / 0B   18.3MB / 4.1kB   6
2023-10-26T14:57:00 fc1ceb3febb8   default   0.01%     83.98MiB / 7.744GiB   1.06%     806B / 0B   18.3MB / 4.1kB   6
2023-10-26T14:57:02 fc1ceb3febb8   default   0.01%     83.98MiB / 7.744GiB   1.06%     806B / 0B   18.3MB / 4.1kB   6
2023-10-26T14:57:04 fc1ceb3febb8   default   0.01%     83.98MiB / 7.744GiB   1.06%     876B / 0B   18.3MB / 4.1kB   6
2023-10-26T14:57:06 fc1ceb3febb8   default   0.01%     83.98MiB / 7.744GiB   1.06%     876B / 0B   18.3MB / 4.1kB   6
2023-10-26T14:57:08 fc1ceb3febb8   default   0.01%     83.98MiB / 7.744GiB   1.06%     876B / 0B   18.3MB / 4.1kB   6
2023-10-26T14:57:10 fc1ceb3febb8   default   0.01%     83.98MiB / 7.744GiB   1.06%     876B / 0B   18.3MB / 4.1kB   6
```

### Last 10 lines

```
root@armadillo-backend:~# tail -10 server-log-script.log 

TIME                CONTAINER ID   NAME      CPU %     MEM USAGE / LIMIT     MEM %     NET I/O          BLOCK I/O        PIDS
2023-10-26T15:19:48 fc1ceb3febb8   default   0.01%     87.44MiB / 7.744GiB   1.10%     59.2kB / 56kB    18.3MB / 4.1kB   5
2023-10-26T15:19:50 29c65e2733ed   rock      0.22%     2.714GiB / 7.744GiB   35.04%    841kB / 179kB    0B / 2.24MB      80
2023-10-26T15:19:50 28fee48632b2   xenon     0.14%     2.319GiB / 7.744GiB   29.94%    4.52MB / 168kB   0B / 4.1kB       65
2023-10-26T15:19:50 fc1ceb3febb8   default   0.01%     87.44MiB / 7.744GiB   1.10%     59.2kB / 56kB    18.3MB / 4.1kB   5
2023-10-26T15:19:52 29c65e2733ed   rock      0.23%     2.714GiB / 7.744GiB   35.04%    841kB / 179kB    0B / 2.24MB      80
2023-10-26T15:19:52 28fee48632b2   xenon     0.11%     2.319GiB / 7.744GiB   29.94%    4.52MB / 168kB   0B / 4.1kB       65
2023-10-26T15:19:52 fc1ceb3febb8   default   0.01%     87.44MiB / 7.744GiB   1.10%     59.2kB / 56kB    18.3MB / 4.1kB   5
2023-10-26T15:19:54 29c65e2733ed   rock      0.23%     2.714GiB / 7.744GiB   35.04%    841kB / 179kB    0B / 2.24MB      80
2023-10-26T15:19:54 28fee48632b2   xenon     0.13%     2.319GiB / 7.744GiB   29.94%    4.52MB / 168kB   0B / 4.1kB       65
2023-10-26T15:19:54 fc1ceb3febb8   default   0.01%     87.44MiB / 7.744GiB   1.10%     59.2kB / 56kB    18.3MB / 4.1kB   5
```

## Analysis of running docker container

```zsh
top
# m to show memory

# apt install htop
htop
# t for tree view

df
du -s -m

# apt install lsof
lsof 
```

**Comment by clemens-tolboom** [comment]
## Rockr

### Kill all sessions

```Rscript
# @see https://github.com/obiba/rockr
library(rockr)

conn <- rockr.connect(username=rock_admin, password=rock_password, url = rock_url)

close_sessions <- function() {
  sessions=rockr.sessions(conn)
  for ( s in sessions) {
    rockr.session_switch(conn, s)
    rockr.close(conn)
  }
}

close_sessions()
```

**Comment by clemens-tolboom** [comment]
## Armadillo

### Session end

I used the Rockr kill all sessions above then stopped Armadillo

```log
16:14:01.562 [SpringApplicationShutdownHook|] DEBUG o.s.web.client.RestTemplate - HTTP DELETE http://localhost:6313/r/session/06083e48-4a62-41df-bcf7-43ebf17b66ae
16:14:01.562 [SpringApplicationShutdownHook|] DEBUG o.s.web.client.RestTemplate - Accept=[application/json, application/*+json]
16:14:01.578 [SpringApplicationShutdownHook|] DEBUG o.s.web.client.RestTemplate - Response 404 NOT_FOUND
16:14:01.579 [SpringApplicationShutdownHook|] WARN  org.molgenis.r.rock.RockConnection - Failure when closing the Rock R session 06083e48-4a62-41df-bcf7-43ebf17b66ae
org.springframework.web.client.HttpClientErrorException$NotFound: 404 Not Found: "{"status":"404","key":"RSessionNotFoundException","args":["06083e48-4a62-41df-bcf7-43ebf17b66ae"],"message":"No such R session: 06083e48-4a62-41df-bcf7-43ebf17b66ae"}"
```

**Comment by clemens-tolboom** [comment]
## Conclusion(?)

The problem seems not to be anywhere but in session management.

That is having 1 session (Data analist) and using it for long times out in 120 minutes

Running running `release-test.R` multiple (N) times creates N new session al to be timeout after 120 minutes.

### Caveat

The PR #522 does add closing the connection which needs closer inspection

**Comment by clemens-tolboom** [comment]
After testing the fix in PR #522 this can be closed

**Comment by clemens-tolboom** [comment]
Reopening as that change is not merged in master.

**Comment by clemens-tolboom** [comment]
Will be closed by PR #534
